numba/array_analysis.py

```diff
@@ -0,0 +1,718 @@
+                    c = self._merge_classes(c, self.array_shape_classes[v.name][i])
+                out_eqs[i] = c
+            # output has one extra dimension
+            new_class1 = self._get_next_class_with_size(len(arr_args))
+            out_eqs.insert(axis, new_class1)
+            # TODO: set size to sum of input array's size along axis
+            return out_eqs
+        elif call_name=='hstack':
+            # hstack is same as concatenate with axis=1 for ndim>=2
+            dummy_one_var = ir.Var(args[0].scope, "__dummy_1", args[0].loc)
+            self.constant_table["__dummy_1"] = 1
+            args.append(dummy_one_var)
+            return self._analyze_np_call('concatenate', args, kws)
+        elif call_name=='dstack':
+            # dstack is same as concatenate with axis=2, atleast_3d args
+            args[0] = self.convert_seq_to_atleast_3d(args[0])
+            dummy_two_var = ir.Var(args[0].scope, "__dummy_2", args[0].loc)
+            self.constant_table["__dummy_2"] = 2
+            args.append(dummy_two_var)
+            return self._analyze_np_call('concatenate', args, kws)
+        elif call_name=='vstack':
+            # vstack is same as concatenate with axis=0 if 2D input dims or more
+            # TODO: set size to sum of input array's size for 1D
+            arr_args = self._get_sequence_arrs(args[0].name)
+            if len(arr_args)==0:
+                return None
+            ndims = self._get_ndims(arr_args[0].name)
+            if ndims>=2:
+                dummy_zero_var = ir.Var(args[0].scope, "__dummy_0", args[0].loc)
+                self.constant_table["__dummy_0"] = 0
+                args.append(dummy_zero_var)
+                return self._analyze_np_call('concatenate', args, kws)
+        elif call_name=='column_stack':
+            # 1D arrays turn into columns of 2D array
+            arr_args = self._get_sequence_arrs(args[0].name)
+            if len(arr_args)==0:
+                return None
+            c = self.array_shape_classes[arr_args[0].name][0]
+            for v in arr_args:
+                c = self._merge_classes(c, self.array_shape_classes[v.name][0])
+            new_class = self._get_next_class_with_size(len(arr_args))
+            return [c , new_class]
+        elif call_name in ['cumsum', 'cumprod']:
+            in_arr = args[0].name
+            in_ndims = self._get_ndims(in_arr)
+            # for 1D, output has same size
+            # TODO: return flattened size for multi-dimensional input
+            if in_ndims==1:
+                return copy.copy(self.array_shape_classes[in_arr])
+        elif call_name=='linspace':
+            # default is 50, arg3 is size
+            LINSPACE_DEFAULT_SIZE = 50
+            size = LINSPACE_DEFAULT_SIZE
+            if len(args)>=3:
+                size = args[2].name
+            new_class = self._get_next_class_with_size(size)
+            return [new_class]
+        elif call_name=='dot':
```

Comment:
Not yet. Handling dot() for 1D args on the to do list in general (for parallelization also).